include/pybind11/numpy.h

```diff
@@ -384,6 +384,73 @@ struct npy_api {
     }
 };
 
+// This table normalizes typenums by mapping NPY_INT_, NPY_LONG, ... to NPY_INT32_, NPY_INT64, ...
+// This is needed to correctly handle situations where multiple typenums map to the same type,
+// e.g. NPY_LONG_ may be equivalent to NPY_INT_ or NPY_LONGLONG_ despite having a different
+// typenum. The normalized typenum always matches the values used in npy_format_descriptor.
+static constexpr int normalized_dtype_num[npy_api::NPY_VOID_ + 1] = {
```

Comment:
Wow ... What if someone tries to make a changes to `enum constants`, could that result in confusing behavior and time-consuming debugging?

Could it make sense to leave "if this than that" comments here and near the top of `enum constants`?

```
// If you change this code, please review `normalized_dtype_num` below.
...
// If you change this code, please review `enum constants` above.
```

Maybe also: Could this new code be moved closer to the `enum constants` code?

Reply:
A bit verbose on the possibility (a `long long` is always 64bit on any supported platform for example), but seems fine to me.
I wonder if "bitsized" or so might be a clearer name than "normalized".

Reply:
> What if someone tries to make a changes to enum constants, could that result in confusing behavior and time-consuming debugging?

Those constants are part of the Numpy API, so that would cause much larger problems. In any case I've added the suggested comment.

> Maybe also: Could this new code be moved closer to the enum constants code?

Do you mean inside `struct npy_api`? That would create some extra complications since it's a static array in a header-only library.

> I wonder if "bitsized" or so might be a clearer name than "normalized".

Do you want to rename just the `normalized_dtype_num` array or also `dtype::normalized_num`?

Reply:
Hm ... I think "normalized" better captures what this is about. What we really want is to normalize (for use in switch statements). That we're picking the bitsize names is a choice of secondary importance.

Reply:
Sounds good, I think normalize is fine.  I agree in the context of switch statements, it is clear enough.